Sources/AWSLambdaRuntimeCore/LambdaContext.swift

```diff
@@ -49,40 +50,118 @@ extension Lambda {
+                self.eventLoop = eventLoop
+                self.allocator = allocator
+            }
+        }
+
+        /// Contains contextual metadata such as request and trace identifiers, along with other information which may
+        /// be carried throughout asynchronous and cross-node boundaries (e.g. through HTTPClient calls).
+        public var baggage: Baggage {
+            get {
+                self.storage.baggage
+            }
+            set {
+                if isKnownUniquelyReferenced(&self.storage) {
+                    self.storage.baggage = newValue
+                } else {
+                    self.storage = _Storage(
+                        baggage: newValue,
+                        invokedFunctionARN: self.storage.invokedFunctionARN,
+                        deadline: self.storage.deadline,
+                        cognitoIdentity: self.storage.cognitoIdentity,
+                        clientContext: self.storage.clientContext,
+                        _logger: self.storage._logger,
+                        eventLoop: self.storage.eventLoop,
+                        allocator: self.storage.allocator
+                    )
+                }
+            }
+        }
+
         /// The request ID, which identifies the request that triggered the function invocation.
-        public let requestID: String
+        public var requestID: String {
+            self.storage.baggage.lambdaRequestID
+        }
 
         /// The AWS X-Ray tracing header.
-        public let traceID: String
+        public var traceID: String {
+            self.storage.baggage.lambdaTraceID
+        }
 
         /// The ARN of the Lambda function, version, or alias that's specified in the invocation.
-        public let invokedFunctionARN: String
+        public var invokedFunctionARN: String {
+            self.storage.invokedFunctionARN
+        }
 
         /// The timestamp that the function times out
-        public let deadline: DispatchWallTime
+        public var deadline: DispatchWallTime {
+            self.storage.deadline
+        }
 
         /// For invocations from the AWS Mobile SDK, data about the Amazon Cognito identity provider.
-        public let cognitoIdentity: String?
+        public var cognitoIdentity: String? {
+            self.storage.cognitoIdentity
+        }
 
         /// For invocations from the AWS Mobile SDK, data about the client application and device.
-        public let clientContext: String?
+        public var clientContext: String? {
+            self.storage.clientContext
+        }
 
-        /// `Logger` to log with
+        /// `Logger` to log with, it is automatically populated with `baggage` information (such as `traceID` and `requestID`).
         ///
         /// - note: The `LogLevel` can be configured using the `LOG_LEVEL` environment variable.
-        public let logger: Logger
+        public var logger: Logger {
+            self.storage._logger.with(self.baggage)
```

Comment:
I can't remember what with() does... does it mean every logger access is a new instance?

Reply:
Yes, every `context.logger` is a "new logger" but a logger is a cheap value type so that's not terrible.

This has to be like that because we need the logger to point to the "latest" baggage values. As everything here has value semantics, we can't just pass a reference to the context or baggage into the handler, and have to pass it when we're about to use it.

The `with` created LogHandler is lazy though -- it does not create Logger.Metadata _yet_ unless a log statement is actually triggered.

Alternatively, this could be done on a `baggage { didSet { ...` if you'd prefer that but I think that's actually hard to implement correctly without growing memory forever... we'd need to notice "aha, this logger already has a baggage logger, so we need to unwrap it and then wrap again with a baggage aware one" to prevent it growing infinitely with `Logger(BaggageLogHandler(BaggageLogHandler(BaggageLogHandler(BaggageLogHandler(BaggageLogHandler(BaggageLogHandler(...` if you know what I mean... This is hard to implement because `handler` is private in logging, so we can't check for it... Implementing it that way may require making the handler public in `SwiftLog`.

Reply:
hmm I think this could actually add overhead for cases that log allot. I think keeping state on when baggage changed otherwise returning the existing logger could be a good optimization. maybe even coded into BaggageContext so users dont have to reinvent

Reply:
I revisited this and will try to implement via an "materialize once" into logging metadata. You're right that this is painful if there's a lot of logging, and we should not make that painful :\ 

It won't be doable as a log handler I guess, so that goes out the window; really I guess what would solve the needless copying is to make Logging depend on Baggage, and then we can remove its storage for metadata and replace it with baggage, avoiding the copying alltogether.

Not sure when to pitch this, it'd be a logging 2.0 I guess?

Reply:
Addressed this -- metadata is now set eagerly and we won't get into crazy copying anymore.

Thanks for reminding me about this, it's kind of obvious in hindsight but I missed it.

Some numbers how the copying "eagerly" amortizes (that's for 10k logs); it's just bad for "never logs" cases but one can assume those don't really happen in a large "actual" system (vs a library) tbh.

The `logger.with` lazy-pattern remains available when one wants to use it instead.

![Screen Shot 2020-09-18 at 18 17 27](https://user-images.githubusercontent.com/120979/93587079-4729c180-f9e4-11ea-8980-285bfc4c405a.png)

Reply:
(orange/red are the `logger.with(baggage)`)